auth/credentials/filetypes.go

```diff
@@ -212,14 +249,34 @@ func handleImpersonatedServiceAccount(f *credsfile.ImpersonatedServiceAccountFil
 	if err != nil {
 		return nil, err
 	}
-	return impersonate.NewTokenProvider(&impersonate.Options{
-		URL:       f.ServiceAccountImpersonationURL,
-		Scopes:    opts.scopes(),
-		Tp:        tp,
-		Delegates: f.Delegates,
-		Client:    opts.client(),
-		Logger:    opts.logger(),
-	})
+	ud := resolveUniverseDomain(opts.UniverseDomain, f.UniverseDomain)
+	impOpts := &impersonate.Options{
+		URL:            f.ServiceAccountImpersonationURL,
+		Scopes:         opts.scopes(),
+		Tp:             tp,
+		Delegates:      f.Delegates,
+		Client:         opts.client(),
+		Logger:         opts.logger(),
+		UniverseDomain: ud,
+	}
+	// For impersonated service accounts, the trust boundary applies to the TARGET service account.
+	// The universe domain should be the one associated with the impersonated account.
+	targetSAEmail, err := impersonate.ExtractServiceAccountEmail(f.ServiceAccountImpersonationURL)
+	if err != nil {
+		return nil, fmt.Errorf("credentials: could not extract target service account email for trust boundary: %w", err)
+	}
+	if trustBoundaryEnabledErr != nil {
+		return nil, trustBoundaryEnabledErr
+	}
+	if trustBoundaryEnabled {
+		targetSATrustBoundaryConfig := trustboundary.NewServiceAccountTrustBoundaryConfig(targetSAEmail, impOpts.UniverseDomain)
+		tbProvider, err := trustboundary.NewTrustBoundaryDataProvider(opts.client(), targetSATrustBoundaryConfig, opts.logger())
+		if err != nil {
+			return nil, fmt.Errorf("credentials: failed to initialize trust boundary provider for impersonation: %w", err)
+		}
+		impOpts.TrustBoundaryDataProvider = &internalDataProviderAdapter{internalProvider: tbProvider}
+	}
```

Comment:
Can the logic here from `ExtractServiceAccountEmail` to setting `impOpts.TrustBoundaryDataProvider` all be moved into `impersonate.NewTokenProvider`, and `impOpts.TrustBoundaryDataProvider` replaced with `impOpts.TrustBoundaryEnabled`? I'm not sure that it can, but if we can encapsulate this logic within the `impersonate` package that would be a big win.